src/Models/Model.php

```diff
@@ -326,10 +327,48 @@ public static function propertyKeys() : array {
 		return array_keys( ( new static() )->properties );
 	}
 
+	/**
+	 * Will infer the type based on the property definition and
+	 * cast the type on the value passed in.
+	 *
+	 * @since TBD
+	 *
+	 * @param string $key   The property name.
+	 * @param mixed  $value The value to be cast.
+	 *
+	 * @return array|bool|int|string The $value after being cast to it's type.
+	 */
+	protected function castAttribute( string $key, $value ) {
+		// Handle nullable fields as expected.
+		if ( is_null( $value ) ) {
+			return null;
+		}
+
+		$type = $this->getPropertyType( $key );
+
+		switch ( $type ) {
+			case 'int':
+				return (int) $value;
+			case 'string':
+				return (string) $value;
+			case 'bool':
+				return (bool) $value;
```

Comment:
Regarding booleans, there is a tricky moment in PHP. I'm not sure if we want to support it here, but we had such case in LearnDash, so want to make sure you are aware.

```php
if ( 'bool' === $property_type ) {
    // We want to convert "false" string to false.
    $property_value = filter_var( $property_value, FILTER_VALIDATE_BOOLEAN );
}